### PR TITLE
Add commit status summary table to reduce query from commit status table

### PR DESCRIPTION
**Pull request by lunny**
This PR adds a new table named commit status summary to reduce queries from the commit status table. After this change, commit status summary table will be used for the final result, commit status table will be for details.

### CODE REVIEW

**Comment by lunny** [comment]
This is ready to be review.

File: models/git/commit_status_summary.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	api "code.gitea.io/gitea/modules/structs"
+
+	"xorm.io/builder"
+)
+
+// CommitStatusSummary holds the latest commit Status of a single Commit
+type CommitStatusSummary struct {
+	ID     int64                 `xorm:"pk autoincr"`
+	RepoID int64                 `xorm:"INDEX UNIQUE(repo_sha_index)"`
+	SHA    string                `xorm:"VARCHAR(64) NOT NULL INDEX UNIQUE(repo_sha_index)"`
```
Comment (wolfogre) [inline models/git/commit_status_summary.go]: Maybe
```suggestion
	RepoID int64                 `xorm:"INDEX UNIQUE(repo_sha)"`
	SHA    string                `xorm:"VARCHAR(64) NOT NULL INDEX UNIQUE(repo_sha)"`
```

File: models/migrations/v1_23/v295.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+package v1_23 //nolint
+
+import "xorm.io/xorm"
+
+func AddCommitStatusSummary(x *xorm.Engine) error {
+	type CommitStatusSummary struct {
+		ID     int64  `xorm:"pk autoincr"`
+		RepoID int64  `xorm:"INDEX UNIQUE(repo_sha_index)"`
+		SHA    string `xorm:"VARCHAR(64) NOT NULL INDEX UNIQUE(repo_sha_index)"`
```
Comment (wolfogre) [inline models/migrations/v1_23/v295.go]: Ditto

```suggestion
		RepoID int64  `xorm:"INDEX UNIQUE(repo_sha)"`
		SHA    string `xorm:"VARCHAR(64) NOT NULL INDEX UNIQUE(repo_sha)"`
```

**Review by wolfogre — COMMENTED** [review]
~~How can the missing `CommitStatusSummary` for the old `CommitStatus` be migrated? I imagine it could be something like "fallback to query `CommitStatus` if there's no `CommitStatusSummary`", but I don't see the code. Did I miss it?~~

Never mind, I see it now.

**Review by wolfogre — APPROVED** [review]

File: models/migrations/v1_23/v295.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+package v1_23 //nolint
+
+import "xorm.io/xorm"
+
+func AddCommitStatusSummary(x *xorm.Engine) error {
+	type CommitStatusSummary struct {
+		ID     int64  `xorm:"pk autoincr"`
+		RepoID int64  `xorm:"INDEX UNIQUE(repo_sha)"`
+		SHA    string `xorm:"VARCHAR(64) NOT NULL INDEX UNIQUE(repo_sha)"`
```
Comment (yp05327) [inline models/migrations/v1_23/v295.go]: repo_id_sha?

File: models/git/commit_status_summary.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+func init() {
+	db.RegisterModel(new(CommitStatusSummary))
+}
+
+type RepoSHA struct {
+	RepoID int64
+	SHA    string
+}
+
+func GetLatestCommitStatusForRepoAndSHAs(ctx context.Context, repoSHAs []RepoSHA) ([]*CommitStatus, error) {
```
Comment (yp05327) [inline models/git/commit_status_summary.go:33]: Just a question: are `[]*RepoSHA` and `[]RepoSHA` different here?

File: models/migrations/v1_23/v295.go (hunk 3 above)
Comment (lunny) [reply models/migrations/v1_23/v295.go]: 32b6790

File: models/git/commit_status_summary.go (hunk 4 above)
Comment (lunny) [reply models/git/commit_status_summary.go:33]: Both are OK in this situation.

**Comment by yp05327** [comment]
git_model.FindBranchesByRepoAndBranchName(ctx, repoBranchNames)
This function name is so strange as the return is commit id not branch
image

File: services/repository/commitstatus/commitstatus.go [hunk 5]
```diff
@@ -114,8 +121,35 @@ func FindReposLastestCommitStatuses(ctx context.Context, repos []*repo_model.Rep
...
+	if err != nil {
+		return nil, fmt.Errorf("GetLatestCommitStatusForRepoAndSHAs: %v", err)
+	}
+
+	for _, summary := range summaryResults {
+		for i, repo := range repos {
+			if repo.ID == summary.RepoID {
+				results[i] = summary
+				_ = slices.DeleteFunc(repoSHAs, func(repoSha git_model.RepoSHA) bool {
+					return repoSha.RepoID == repo.ID
```
Comment (yp05327) [inline services/repository/commitstatus/commitstatus.go]: repoSHA?

Comment (lunny) [reply services/repository/commitstatus/commitstatus.go]: 027ae02349

File: models/migrations/migrations.go [hunk 6]
```diff
@@ -575,8 +575,10 @@ var migrations = []Migration{
 	NewMigration("Ensure every project has exactly one default column", v1_22.CheckProjectColumnsConsistency),
 
 	// Gitea 1.22.0 ends at 294
-
```
Comment (yp05327) [inline models/migrations/migrations.go:578]: no delete line

**Review by yp05327 — APPROVED** [review]

**Comment by yp05327** [comment]
There's another issue maybe come from #29444. Should we fix it in this PR or another one?
`git_model.CommitStatus.TargetURL` is necessary in `FindReposLastestCommitStatuses` or the link will broken.
image
image

**Comment by lunny** [comment]
> There's another issue maybe come from #29444. Should we fix it in this PR or another one? `git_model.CommitStatus.TargetURL` is necessary in `FindReposLastestCommitStatuses` or the link will broken. image image

It's a cache implementation bug and I will fix it in another PR so that it can be backport easier.

File: models/migrations/migrations.go (hunk 6 above)
Comment (lunny) [reply models/migrations/migrations.go:578]: d46bdd1c7d

**Comment by yp05327** [comment]
I see, the cache only caches status.

**Review by yp05327 — APPROVED** [review]

**Pull request merged**